### PR TITLE
Elbv3

### PR DESCRIPTION
**Pull request by anton-sidelnikov**
- [x] az
- [x] quota
- [x] certificate
- [x] listener
- [x] load_balancer
- [x] flavor

### CODE REVIEW

File: otcextensions/sdk/vlb/v3/availability_zone.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            raise exceptions.MethodNotSupported(cls, "list")
+        session = cls._get_session(session)
+        microversion = cls._get_microversion_for_list(session)
+
+        if base_path is None:
+            base_path = cls.base_path
+
+        response = session.get(
+            base_path,
+            headers={"Accept": "application/json"},
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/availability_zone.py:43]: is this the only need to override function? If yes then we have other ways to inject headers (https://github.com/opentelekomcloud/python-otcextensions/blob/master/otcextensions/sdk/rds/v3/_proxy.py#L30)

File: otcextensions/sdk/vlb/v3/availability_zone.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+class AvailabilityZone(resource.Resource):
+    base_path = '/elb/availability-zones'
+
+    # capabilities
+    allow_list = True
+
+    # Properties
+    #: Specifies the AZ status.
+    state = resource.Body('state', type=str)
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/availability_zone.py]: please sort attrs

File: otcextensions/sdk/vlb/v3/flavor.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    allow_list = True
+
+    _query_mapping = resource.QueryParameters(
+        'marker', 'limit', 'page_reverse',
+        'id', 'name', 'type', 'shared'
+    )
+
+    # Properties
+    #: Specifies the flavor ID.
+    id = resource.Body('id')
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/flavor.py:32]: please sort attrs

File: otcextensions/sdk/vlb/v3/load_balancer.py [hunk 4]
```diff
@@ -29,20 +29,37 @@ class LoadBalancer(resource.Resource):
         'description', 'flavor_id', 'name', 'project_id', 'provider',
         'vip_address', 'vip_network_id', 'vip_port_id', 'vip_subnet_id',
         'vip_qos_policy_id', 'provisioning_status', 'operating_status',
-        'availability_zone', is_admin_state_up='admin_state_up',
+        'availability_zone_list', is_admin_state_up='admin_state_up',
+        availability_zone='availability_zone_list',
     )
 
     # Properties
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/load_balancer.py:39]: list is really long. Maintenance would be complex if we do not sort it

File: otcextensions/sdk/vlb/v3/load_balancer.py [hunk 5]
```diff
@@ -29,20 +29,37 @@ class LoadBalancer(resource.Resource):
...
-    #: The administrative state of the load balancer *Type: bool*
-    is_admin_state_up = resource.Body('admin_state_up', type=bool)
     #: Name of the target Octavia availability zone
-    availability_zone = resource.Body('availability_zone', type=list)
+    availability_zone = resource.Body('availability_zone_list', type=list)
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/load_balancer.py]: what if we do `availability_zones`. We make it easier not to force list2string conversion where we require to normalize things

File: otcextensions/sdk/vlb/v3/load_balancer.py [hunk 6]
```diff
@@ -29,20 +29,37 @@ class LoadBalancer(resource.Resource):
...
     #: The load balancer description
     description = resource.Body('description')
+    #: EIP bound to the load balancer.
+    eips = resource.Body('eips', type=list, list_type=dict)
     #: The load balancer flavor ID
     flavor_id = resource.Body('flavor_id')
+    #: Specifies whether the load balancer is a dedicated load balancer.
+    guaranteed = resource.Body('guaranteed')
+    #: Load balancer ID.
+    id = resource.Body('id')
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/load_balancer.py]: id/name are general attrs in the base class and they do not need to be explicitly included

File: otcextensions/sdk/vlb/v3/quota.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class Quota(resource.Resource):
+    base_path = '/elb/quotas'
+
+    # capabilities
+    allow_fetch = True
+
+    # Properties
+    #: Specifies the certificate quota.
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/quota.py:23]: please also sort here

File: otcextensions/sdk/vlb/v3/quota.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    members_per_pool = resource.Body('members_per_pool', type=int)
+    #: Specifies the backend server group quota.
+    pool = resource.Body('pool', type=int)
+    #: Specifies the IP address group quota.
+    ipgroup = resource.Body('ipgroup', type=int)
+    #: Specifies the IP address group quota.
+    project_id = resource.Body('project_id', type=str)
+
+    @classmethod
+    def get(cls, session):
```
Comment (gtema) [inline otcextensions/sdk/vlb/v3/quota.py:45]: look at https://opendev.org/openstack/openstacksdk/src/branch/master/openstack/block_storage/v3/_proxy.py#L525 This is a very similar case where we have only "GET" without ID for the resource. You do not need to override `get`

File: otcextensions/tests/functional/sdk/vlb/v3/test_loadbalancer.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        # self.vlb = self.client.create_load_balancer(
+        #     name=self.vlb_name
+        # )
+        #
+        # self.addCleanup(self.conn.vlb.delete_load_balancer, self.vlb)
+
+
+    def test_create_loadbalancer(self):
+        # neutron_subnet_id = self.net_client.subnets()
+        elb = self.client.create_load_balancer(**self.lb_attrs_eip)
```
Comment (gtema) [inline otcextensions/tests/functional/sdk/vlb/v3/test_loadbalancer.py]: don't forget to cleanup

File: otcextensions/tests/functional/sdk/vlb/v3/test_quota.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+from otcextensions.tests.functional.sdk.vlb import TestVlb
+
+
+class TestQuota(TestVlb):
+
+    def setUp(self):
+        super(TestQuota, self).setUp()
+
+    def test_get_quotas(self):
+        qt = self.client.quotas()
```
Comment (gtema) [inline otcextensions/tests/functional/sdk/vlb/v3/test_quota.py]: as commented above - the interface may be other `self.client.get_quotas`

**Review by gtema — COMMENTED** [review]
you also would need to add unit tests for the resources

File: otcextensions/sdk/vlb/v3/availability_zone.py (hunk 1 above)
Comment (anton-sidelnikov) [reply otcextensions/sdk/vlb/v3/availability_zone.py:43]: not only, also this "for raw_resource in resources[0]:"

File: otcextensions/sdk/vlb/v3/quota.py (hunk 8 above)
Comment (anton-sidelnikov) [reply otcextensions/sdk/vlb/v3/quota.py:45]: this is not fit here, maybe i just need to override _translate_response, this is the place where i loose response info

**Comment by gtema** [comment]
recheck

**Pull request merged**